### PR TITLE
add endpoint for getting auth types

### PR DESCRIPTION
**Pull request by bradsawadye**
This endpoint is for getting the enabled authentication types

TRACE-128

### CODE REVIEW

File: src/api/authentication.js [hunk 1]
```diff
@@ -217,5 +225,42 @@ export async function authenticate (ctx, next) {
...
 }
 
+export async function getEnabledAuthenticationTypes (ctx, next) {
+  // Log level for messages logged
+  let logLevel = 'error'
+
+  try {
+    if (!authorisation.inGroup('admin', ctx.authenticated)) {
+      ctx.statusCode = 403
+      const error = `User ${ctx.authenticated.email} is not an admin, API access to get enabled authentication types denied.`
```
Comment (MattyJ007) [inline src/api/authentication.js:235]: @bradsawadye This const here is unnecessary as it's only used once.

File: src/api/authentication.js [hunk 2]
```diff
@@ -217,5 +225,42 @@ export async function authenticate (ctx, next) {
...
+      logLevel = 'info'
+      throw Error(error)
```
Comment (MattyJ007) [inline src/api/authentication.js:237]: I know that `throw Error()` and `throw new Error()` are equivalent, but for project consistency could you please change this to `new Error()` (or if you really want, change all the existing `new Error()`s to `Error()`).
I do like the efficiency of just `Error()` but whatever you choose we need to keep one style throughout the repo

File: src/constants.js [hunk 3]
```diff
@@ -1,3 +1,11 @@
...
 
 export const DEFAULT_API_PORT = 8080
+
+export const MUTUAL_TLS_AUTH_TYPE = 'mutual-tls-auth'
+
+export const BASIC_AUTH_TYPE = 'basic-auth'
+
+export const CUSTOM_TOKEN_AUTH_TYPE = 'custom-token-auth'
+
+export const JWT_TOKEN_AUTH_TYPE = 'jwt-token-auth'
```
Comment (MattyJ007) [inline src/constants.js:11]: @bradsawadye technically you've just said `JSON Web Token token auth`... So you can just leave it at `jwt-auth`

File: src/api/authentication.js [hunk 4]
```diff
@@ -217,5 +225,42 @@ export async function authenticate (ctx, next) {
...
+    }
+
+    if (
+      !config.authentication ||
+      !Object.keys(config.authentication).length
+    ) {
+      throw Error('Invalid authentication Types, openhim not configured correctly')
```
Comment (MattyJ007) [inline src/api/authentication.js:244]: The wording here isn't quite right. Maybe something like below

```suggestion
      throw Error('No authentication enabled, invalid OpenHIM configuration')
```

File: test/integration/authenticationAPITests.js [hunk 5]
```diff
@@ -20,6 +20,58 @@ import { config } from '../../src/config'
...
+      await AuditModel.deleteMany({})
+    })
+
+    after(async () => {
+      await Promise.all([
+        testUtils.cleanupTestUsers(),
+        promisify(server.stop)()
+      ])
+    })
+
```
Comment (MattyJ007) [inline test/integration/authenticationAPITests.js:45]: @bradsawadye Please add a test case for when the no authentication is configured.

File: src/api/authentication.js [hunk 6]
```diff
@@ -217,5 +225,42 @@ export async function authenticate (ctx, next) {
   return next()
 }
 
+export async function getEnabledAuthenticationTypes (ctx, next) {
+  // Log level for messages logged
+  let logLevel = 'error'
+
+  try {
```
Comment (MattyJ007) [inline src/api/authentication.js:232]: @bradsawadye Why did you surround everything in a `try-catch`?

Comment (bradsawadye) [reply src/api/authentication.js]: addressed in https://github.com/jembi/openhim-core-js/pull/1082/commits/75107c24ed1e86da1944b25de6bf4ff4368c007a

File: test/integration/authenticationAPITests.js (hunk 5 above)
Comment (bradsawadye) [reply test/integration/authenticationAPITests.js:50]: addressed in https://github.com/jembi/openhim-core-js/pull/1082/commits/562dc5c4c469b99a04afcc0d09a9d002436aac15

File: src/constants.js (hunk 3 above)
Comment (bradsawadye) [reply src/constants.js]: addressed in https://github.com/jembi/openhim-core-js/pull/1082/commits/ec811918f9e9bfb287e433282d3c8b526563de66

File: src/api/authentication.js (hunk 1 above)
Comment (bradsawadye) [reply src/api/authentication.js]: addressed in https://github.com/jembi/openhim-core-js/pull/1082/commits/562dc5c4c469b99a04afcc0d09a9d002436aac15

File: src/api/authentication.js (hunk 2 above)
Comment (bradsawadye) [reply src/api/authentication.js]: addressed in https://github.com/jembi/openhim-core-js/pull/1082/commits/562dc5c4c469b99a04afcc0d09a9d002436aac15

File: src/api/authentication.js [hunk 7]
```diff
@@ -7,12 +7,20 @@ import logger from 'winston'
...
+import * as authorisation from './authorisation'
 import { UserModelAPI } from '../model/users'
-import { caseInsensitiveRegex } from '../utils'
+import { caseInsensitiveRegex, logAndSetResponse } from '../utils'
 import { config } from '../config'
+import {
+  MUTUAL_TLS_AUTH_TYPE,
+  BASIC_AUTH_TYPE,
+  CUSTOM_TOKEN_AUTH_TYPE,
+  JWT_TOKEN_AUTH_TYPE
```
Comment (MattyJ007) [inline src/api/authentication.js:18]: @bradsawadye You updated this variable name in constants

Comment (MattyJ007) [reply src/api/authentication.js:18]: Also, minor, but can you order the constants import variables alphabetically

File: src/constants.js [hunk 8]
```diff
@@ -1,3 +1,11 @@
...
 
 export const DEFAULT_API_PORT = 8080
+
+export const MUTUAL_TLS_AUTH_TYPE = 'mutual-tls-auth'
+
+export const BASIC_AUTH_TYPE = 'basic-auth'
+
+export const CUSTOM_TOKEN_AUTH_TYPE = 'custom-token-auth'
+
+export const JWT_AUTH_TYPE = 'jwt-token-auth'
```
Comment (MattyJ007) [inline src/constants.js:11]: @bradsawadye you still have `jwt-token`

File: test/integration/authenticationAPITests.js [hunk 9]
```diff
@@ -20,6 +20,75 @@ import { config } from '../../src/config'
...
+        .set('auth-username', testUtils.rootUser.email)
+        .set('auth-ts', authDetails.authTS)
+        .set('auth-salt', authDetails.authSalt)
+        .set('auth-token', authDetails.authToken)
+        .expect(500)
+    })
+
+    it('should retrieve enabled authentication types', async () => {
+      config.authentication.enableMutualTLSAuthentication = true
+      config.authentication.enableBasicAuthentication = false
```
Comment (MattyJ007) [inline test/integration/authenticationAPITests.js:75]: @bradsawadye these current tests didn't pick up that the constant import declaration was incorrect. Can you also test the JWT and Basic auth

**Pull request merged**